### PR TITLE
K ldiv branch

### PR DESCRIPTION
**Pull request by Nisone2000**
I changed the regularizer and made some changes to include cluster centers in the optimizer.

### CODE REVIEW

File: sensorium/models/models.py [hunk 1]
```diff
@@ -28,11 +28,13 @@ def stacked_core_full_gauss_readout(
     hidden_dilation=1,
     laplace_padding=None,
     input_regularizer="LaplaceL2norm",
+    regularizer_type="adaptive_log_norm",
     use_avg_reg=False,
     init_mu_range=0.2,
     init_sigma=1.0,
     readout_bias=True,
     gamma_readout=4,
+    feature_reg_weight=10,
```
Comment (pollytur) [inline sensorium/models/models.py:37]: @Nisone2000  I would still set these 2 for the same number to avoid confusion

File: sensorium/training/trainers.py [hunk 2]
```diff
@@ -36,7 +36,7 @@ def standard_trainer(
     maximize=True,
     tolerance=1e-6,
     restore_best=True,
-    lr_decay_steps=3,
+    lr_decay_steps=5,
```
Comment (pollytur) [inline sensorium/training/trainers.py:39]: @Nisone2000 why?

File: sensorium/training/trainers.py [hunk 3]
```diff
@@ -274,22 +296,29 @@ def full_objective(model, dataloader, data_key, *args, **kwargs):
                     features = readout.features.cpu().detach().squeeze().T.numpy()
```
Comment (pollytur) [inline sensorium/training/trainers.py:304]: @Nisone2000 why do you do detach here? It would stop gradients through the features for clustering loss. Is it only for the initialization purposes?

File: sensorium/training/trainers.py [hunk 4]
```diff
@@ -274,22 +296,29 @@ def full_objective(model, dataloader, data_key, *args, **kwargs):
...
                 # _, accuracy = cluster_accuracy(predicted, actual.cpu().numpy())
-                cluster_centers = torch.tensor(
-                    kmeans.cluster_centers_, dtype=torch.float, requires_grad=True
-                )
-                cluster_centers = cluster_centers.to(device, non_blocking=True)
+            cluster_centers = torch.tensor(
+                kmeans.cluster_centers_, dtype=torch.float
+            )  # Convert to tensor
+            cluster_centers = cluster_centers.to(
+                device, non_blocking=True
```
Comment (pollytur) [inline sensorium/training/trainers.py]: @Nisone2000 what is `non_blocking=True` does?

File: sensorium/training/trainers.py [hunk 5]
```diff
@@ -334,34 +364,39 @@ def full_objective(model, dataloader, data_key, *args, **kwargs):
...
 
-                    features_subset = torch.cat(features_subset, dim=1)
+                    # features_subset = torch.cat(features_subset, dim=1)
                     feature_list = torch.cat(feature_list, dim=1)
                     output = soft_assignments(feature_list, cluster_centers)
-                    target = target_distribution(output)
+                    # detach targets to treet them as pseudolabels for clusters
+                    target = target_distribution(output).detach()
 
                     # To avoid underflow issues when computing this quantity, this loss expects the argument input in the log-space.
```
Comment (pollytur) [inline sensorium/training/trainers.py:387]: nice, thanks for the comments like this!

File: sensorium/training/trainers.py (hunk 3 above)
Comment (Nisone2000) [reply sensorium/training/trainers.py:304]: Yes that's just for the initialization. I also have with torch.no_grad() around this so it's not necessary anymore. I can remove it.

File: sensorium/training/trainers.py (hunk 2 above)
Comment (Nisone2000) [reply sensorium/training/trainers.py:39]: I did this when I had the issue with the peak for the cluster centers still with the orientation equivariant model to get better results. Now I don't really need it anymore.

File: sensorium/training/trainers.py (hunk 4 above)
Comment (Nisone2000) [reply sensorium/training/trainers.py]: This is an old version. I don't know why you see it right there. When I look at the trainer.py file I have this:
            cluster_centers.data = torch.tensor(
                    kmeans.cluster_centers_, dtype=torch.float, device=device
            )

**Comment by pollytur** [comment]
@Nisone2000 what is going with this pull request? should it be merged to main or sth should still be fixed?
